### PR TITLE
GQA for Sentinel-2

### PR DESCRIPTION
**Pull request by uchchwhash**
A barely working version of GQA for Sentinel-2.
Breaks everything Landsat. Is not modular. Duplicate code everywhere.

### CODE REVIEW

File: eugl/gqa/tasks.py [hunk 1]
```diff
@@ -55,17 +71,455 @@
...
+            # failed because GQA cannot be calculated
+            _write_failure_yaml(temp_yaml, self.granule, str(ve))
+            with open(pjoin(temp_directory, 'gverify.log'), 'w') as src:
+                src.write('gverify was not executed because:\n')
+                src.write(str(ve))
+
+        self.output().makedirs()
+        shutil.copy(temp_yaml, self.output().path)
+
+        for temp_log in glob.glob(pjoin(temp_directory, '*gverify.log')):
```
Comment (ASVincent) [inline eugl/gqa/tasks.py]: Nitpick: Can you add a comment to specify the intent of this code.
My interpretation is "If a gverify log exists move it outside of the nested work directory"

To me the for-each break after first language is imprecise and a comment will help clarify intent.

File: eugl/gqa/tasks.py [hunk 2]
```diff
@@ -55,17 +71,455 @@
...
+    # Compute new values to refine the selection
+    for i in range(iterations):
+        # Look for any residuals
+        subset = subset[(abs(subset.X_Residual - current['mean']['x']) < (stddev * current['stddev']['x'])) &
+                        (abs(subset.Y_Residual - current['mean']['y']) < (stddev * current['stddev']['y']))]
+
+        # Re-calculate the mean and standard deviation for both X & Y residuals
+        current = calculate_stats(subset)
+
+        # Calculate abs mean value for both X & Y residuals
```
Comment (ASVincent) [inline eugl/gqa/tasks.py]: Delete this floating comment?

File: eugl/gqa/tasks.py [hunk 3]
```diff
@@ -55,17 +71,455 @@
...
-CONFIG.add_config_path(resource_filename('gqa', 'gqa-defaults.luigi.cfg'))
+
+# TODO purge this
+# CONFIG.add_config_path(resource_filename('gqa', 'gqa-defaults.luigi.cfg'))
 
 # TODO variable change; scene id is landsat specific, level1 is more generic
 # TODO remove refs to l1t (landsat specific)
 
 
+class GQATask(luigi.Task):
```
Comment (ASVincent) [inline eugl/gqa/tasks.py:78]: Can we add a comment that this task is only useful with the singlefile workflow in wagl; (determined by the DataStandardisation dependency).

File: eugl/gqa/tasks.py [hunk 4]
```diff
@@ -55,17 +71,455 @@
 # rather than find and read the internal cfg.
 # Similar to how wagl and tesp do it
 CONFIG = luigi.configuration.get_config()
-CONFIG.add_config_path(resource_filename('gqa', 'gqa-defaults.luigi.cfg'))
+
+# TODO purge this
+# CONFIG.add_config_path(resource_filename('gqa', 'gqa-defaults.luigi.cfg'))
```
Comment (ASVincent) [inline eugl/gqa/tasks.py]: Is this purgeable now or later?

File: eugl/gqa/tasks.py [hunk 5]
```diff
@@ -4,38 +4,53 @@
...
 import os
-from os.path import join as pjoin, dirname, basename, exists, isdir
+from os.path import join as pjoin, dirname, basename, exists, isdir, splitext, abspath
 import glob
 from pkg_resources import resource_filename
 import shutil
 import traceback
 import argparse
+from datetime import datetime
+from dateutil.tz import tzutc
```
Comment (ASVincent) [inline eugl/gqa/tasks.py]: Nitpick: the utc timezone is also available in datetime.timezone(.utc); if you use this it removes the dependency on dateutil from this file.

**Review by ASVincent — APPROVED** [review]
On the whole looks good :+1:

File: eugl/gqa/tasks.py (hunk 1 above)
Comment (uchchwhash) [reply eugl/gqa/tasks.py]: Changed this to which probably is clearer in intent:

```python
        temp_log = glob.glob(pjoin(temp_directory, '*gverify.log'))[0]
        shutil.copy(temp_log, pjoin(self.workdir, basename(temp_log)))
```

Fails if there is no such `.log` files.

File: eugl/gqa/tasks.py (hunk 2 above)
Comment (uchchwhash) [reply eugl/gqa/tasks.py]: Yes!

File: eugl/gqa/tasks.py (hunk 3 above)
Comment (uchchwhash) [reply eugl/gqa/tasks.py:78]: Done.

File: eugl/gqa/tasks.py (hunk 4 above)
Comment (uchchwhash) [reply eugl/gqa/tasks.py]: Since we are going to refactor it to work with Landsat anyway, now. Done.

File: eugl/gqa/tasks.py (hunk 5 above)
Comment (uchchwhash) [reply eugl/gqa/tasks.py]: Done.

**Review by sixy6e — APPROVED** [review]
Yeah it looks fine to me.  Landsat re-integration can occur later (might come sooner than we think), and code niceties are less important than simply having some functionality.

**Pull request merged**